### PR TITLE
Remove QueryShardContext#getMapperService

### PR DESCRIPTION
**Pull request by javanna**
We have been removing usages of getMapperService with the purpose of removing the method at some point. This is so we don't have to expose the entire MapperService to all of the users of QueryShardContext, but rather only specific methods so we can better control its usages, especially around resolving field types.

This commit removes the last few usages and adds the missing methods to QueryShardContext. They all revolve around retrieving DocumentMapper from MapperService, yet we can expose specific methods without even exposing DocumentMapper, which would also cause issues.

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/index/query/QueryShardContext.java [hunk 1]
```diff
@@ -216,6 +221,23 @@ public void addNamedQuery(String name, Query query) {
...
+        return documentMapper == null ? null : documentMapper.mappers().indexAnalyzer();
+    }
+
+    public boolean hasNested() {
+        return mapperService.hasNested();
+    }
+
+    public boolean hasMappings() {
+        return mapperService.documentMapper() != null;
+    }
```
Comment (romseygeek) [inline server/src/main/java/org/elasticsearch/index/query/QueryShardContext.java:239]: This is only used for the RandomScoreFunctionBuilder, correct?  Do we need to add it, or can we just remove that 'no mappings' special case?

Comment (javanna) [reply server/src/main/java/org/elasticsearch/index/query/QueryShardContext.java:239]: yes that is surprisingly the only usage. but I don't feel comfortable removing that no mappings special case. I think it's legit, although annoying to support, and we test it.

**Comment by javanna** [comment]
run elasticsearch-ci/1

**Review by nik9000 — APPROVED** [review]
Awesome!

File: server/src/main/java/org/elasticsearch/index/query/QueryShardContext.java (hunk 1 above)
Comment (javanna) [reply server/src/main/java/org/elasticsearch/index/query/QueryShardContext.java:239]: @romseygeek are you ok with this or maybe do you have other ideas?

Comment (romseygeek) [reply server/src/main/java/org/elasticsearch/index/query/QueryShardContext.java:239]: Let's leave it in for now, if we decide to go with a no-op mapper by default in MapperService we can remove it then.

Comment (javanna) [reply server/src/main/java/org/elasticsearch/index/query/QueryShardContext.java:239]: true, I will try that idea very soon.

**Pull request merged**